test/unit/org/apache/cassandra/auth/AuthCacheInvalidationTest.java

```diff
@@ -63,146 +83,161 @@ public void prepare() throws Throwable
 
-        DatabaseDescriptor.getAuthManager().invalidateCaches();
+        Roles.clearCache();
+        AuthenticatedUser.clearCache();
+    }
+
+    private void createRole(RoleResource role) throws Throwable
+    {
+        executeNet(format("CREATE ROLE %s WITH LOGIN = true AND password = '%s'", role.getRoleName(), userPw));
     }
 
     @Test
     public void testInvalidationOnGrant() throws Throwable
     {
+        populateAndCheckCaches();
+        
         // Checks that the user has no SELECT permissions on tab2
-
-        useUser(user1.getRoleName(), userPw);
-
-        assertUnauthorized("SELECT * FROM " + ks + '.' + tab2,
-                           "User " + user1.getRoleName() + " has no SELECT permission on <table " + ks + '.' + tab2 + "> or any of its parents");
+        assertSelectNotAllowed(user1, ks, tab2);
+        assertSelectNotAllowed(role1, ks, tab2);
+        assertSelectNotAllowed(role2, ks, tab2);
 
         // Checks that the caches have been populated by the query.
         assertRolesCacheContains(user1, setOf(role1));
         assertRolesCacheContains(role1, setOf(role2));
         assertRolesCacheContains(role2, setOf());
 
-        assertPermissionsCacheContains(user1, emptyMap());
-        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, UPDATE));
-        assertPermissionsCacheContains(role2, permissionPerResource(DataResource.keyspace(ks), UPDATE));
+        assertPermissionsCacheContains(user1, emptyPermissionPerResource(DataResource.table(ks, tab2)));
+        assertPermissionsCacheContains(user1, permissionPerResource(DataResource.table(ks, tab), SELECT, MODIFY));
+        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, MODIFY));
+        assertPermissionsCacheContains(role2, permissionPerResource(DataResource.keyspace(ks), MODIFY));
 
         // Allow the user to execute SELECT on tab2
         useSuperUser();
         executeNet("GRANT SELECT ON KEYSPACE " + ks + " TO " + role2.getRoleName());
 
-        // Checks that only the role2 data have been removed from the caches
-        assertRolesCacheContains(user1, setOf(role1));
-        assertRolesCacheContains(role1, setOf(role2));
+        // Checks that all role data have been removed from the caches
+        assertRolesCacheDoesNotContains(user1);
+        assertRolesCacheDoesNotContains(role1);
         assertRolesCacheDoesNotContains(role2);
 
-        assertPermissionsCacheContains(user1, emptyMap());
-        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, UPDATE));
+        assertPermissionsCacheDoesNotContains(user1, DataResource.keyspace(ks));
+        assertPermissionsCacheDoesNotContains(user1, DataResource.table(ks, tab2));
+        assertPermissionsCacheContains(role1, permissionPerResource(DataResource.table(ks, tab), SELECT, MODIFY));
```

Comment:
Since keysapce `ks` is the target of the `GRANT`, I included a check that user1 does not have a perm cache entry for the keyspace as well as tab2.

Reply:
Note that some inconsistency developed as I worked on these tests; in the previous test it used
```
assertPermissionsCacheContains(user1, emptyPermissionPerResource(DataResource.table(ks, tab2)));
```
and here I opted for
```
assertPermissionsCacheDoesNotContains(user1, DataResource.table(ks, tab2));
```
The should be equivalent - they check that either the permission cache key `<user,IResource>` does not exist or maps to an empty set.

I think the `assertPermissionsCacheDoesNotContains` reads a little better.